[wallet] Set the correct nSequence when RBF and OP_CSV are used

afilini opened the pull request:
This commit also fixes the timelock comparing logic in the policy module, since the rules are different for absolute (OP_CLTV) and relative (OP_CSV) timelocks.

Fixes #215

Still in draft since tests are missing.

Also, the big `match` to determine the `nSequence` value is horrible, if anybody has any ideas to improve it I'm open to suggestions.

afilini committed:
[wallet] Set the correct nSequence when RBF and OP_CSV are used

This commit also fixes the timelock comparing logic in the policy module, since
the rules are different for absolute (OP_CLTV) and relative (OP_CSV) timelocks.

Fixes #215

RCasatta commented:
I still need to grasp all the nuances, however, an idea to simplify the `match` could be to introduce the `RBF` type that eagerly fails if MSB is set and maybe have `check_nsequence_rbf` implemented on it.

Maybe `check_` functions could return `Result` and be called before the `match`?

Maybe makes sense to make dedicated structs also for locktime and csv?

afilini committed:
[wallet] Add tests for `check_nsequence_rbf` and `check_nlocktime`

afilini commented:
I've added a few more tests for the two functions that contain the logic to compare nsequence and nlocktime values. According to a code coverage report I've generated locally they should both be 100% covered, as well as all the `match` branches.

I've also made some improvements to the big `match` I was talking about, specifically I removed the `(usize, bool)` tuple with a proper enum type and reordered and documented the branches a little bit. I think it's definitely more readable now, even though there could be more room for improvements later on.

Though, at the moment I would try to merge this as-is, considering the fact that this is a bugfix and we have the release coming soon..

RCasatta commented:
RBFValue make things a little clearer,  it may implement Default one day (not because shorter than RBFValue::Default but so you can use in std methods like `unwrap_or_default()`) however I don't want to block this bugfix we can improve in future PRs

utack eadf50042cd6ffde3eddc5ba1cd2bc3082850b73

afilini commented:
I had done that initially, but then I realized that `None` actually has a different meaning than `Some(RBFValue::Default)`: `None` means no rbf, while the other means to enable rbf with the "default" nsequence value
